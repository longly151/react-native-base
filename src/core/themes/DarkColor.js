module.exports = {
  dark: true,
  colors: {
    primary: '#219EE5',
    background: '#000000',
    card: '#1c1c1e',
    text: '#ffffff',
    border: '#272729',
    notification: '#ff453a',
    grey5: '#393e42',
    grey4: '#43484d',
    grey3: '#5e6977',
    grey2: '#86939e',
    grey1: '#bdc6cf',
    grey0: '#e1e8ee',
    white: '#080808',
    black: '#f2f2f2',
    info: '#36A3F7',
    success: '#34BFA3',
    warning: '#FFB822',
    error: '#F4516C',
  },
};
