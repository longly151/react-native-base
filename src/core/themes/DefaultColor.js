module.exports = {
  dark: false,
  colors: {
    primary: '#219EE5',
    background: '#f2f2f7',
    card: '#ffffff',
    text: '#000000',
    border: '#d8d8d8',
    notification: '#ff3b30',
    grey0: '#393e42',
    grey1: '#43484d',
    grey2: '#5e6977',
    grey3: '#86939e',
    grey4: '#bdc6cf',
    grey5: '#e1e8ee',
    white: '#ffffff',
    black: '#242424',
    info: '#36A3F7',
    success: '#34BFA3',
    warning: '#FFB822',
    error: '#F4516C',
  },
};
