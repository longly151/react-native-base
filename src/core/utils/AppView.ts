import { Font } from '@fonts';
import DefaultColor from '@themes/DefaultColor';
import { Dimensions, StyleProp, ViewStyle, TextStyle, ColorValue, ScaledSize } from 'react-native';
import ViewConfig from '@themes/ViewConfig';

const { width, height } = Dimensions.get('window');

/**
 * Shadow
 */
const androidDepth = {
  umbra: [
    '0px 2px 1px -1px',
    '0px 3px 1px -2px',
    '0px 3px 3px -2px',
    '0px 2px 4px -1px',
    '0px 3px 5px -1px',
    '0px 3px 5px -1px',
    '0px 4px 5px -2px',
    '0px 5px 5px -3px',
    '0px 5px 6px -3px',
    '0px 6px 6px -3px',
    '0px 6px 7px -4px',
    '0px 7px 8px -4px',
    '0px 7px 8px -4px',
    '0px 7px 9px -4px',
    '0px 8px 9px -5px',
    '0px 8px 10px -5px',
    '0px 8px 11px -5px',
    '0px 9px 11px -5px',
    '0px 9px 12px -6px',
    '0px 10px 13px -6px',
    '0px 10px 13px -6px',
    '0px 10px 14px -6px',
    '0px 11px 14px -7px',
    '0px 11px 15px -7px',
  ],
  penumbra: [
    '0px 1px 1px 0px',
    '0px 2px 2px 0px',
    '0px 3px 4px 0px',
    '0px 4px 5px 0px',
    '0px 5px 8px 0px',
    '0px 6px 10px 0px',
    '0px 7px 10px 1px',
    '0px 8px 10px 1px',
    '0px 9px 12px 1px',
    '0px 10px 14px 1px',
    '0px 11px 15px 1px',
    '0px 12px 17px 2px',
    '0px 13px 19px 2px',
    '0px 14px 21px 2px',
    '0px 15px 22px 2px',
    '0px 16px 24px 2px',
    '0px 17px 26px 2px',
    '0px 18px 28px 2px',
    '0px 19px 29px 2px',
    '0px 20px 31px 3px',
    '0px 21px 33px 3px',
    '0px 22px 35px 3px',
    '0px 23px 36px 3px',
    '0px 24px 38px 3px',
  ],
  ambient: [
    '0px 0px 0px 0px',
    '0px 1px 3px 0px',
    '0px 1px 5px 0px',
    '0px 1px 8px 0px',
    '0px 1px 10px 0px',
    '0px 1px 14px 0px',
    '0px 1px 18px 0px',
    '0px 2px 16px 1px',
    '0px 3px 14px 2px',
    '0px 3px 16px 2px',
    '0px 4px 18px 3px',
    '0px 4px 20px 3px',
    '0px 5px 22px 4px',
    '0px 5px 24px 4px',
    '0px 5px 26px 4px',
    '0px 6px 28px 5px',
    '0px 6px 30px 5px',
    '0px 6px 32px 5px',
    '0px 7px 34px 6px',
    '0px 7px 36px 6px',
    '0px 8px 38px 7px',
    '0px 8px 40px 7px',
    '0px 8px 42px 7px',
    '0px 9px 44px 8px',
    '0px 9px 46px 8px',
  ],
};
function interpolate(i: number, a: number, b: number, a2: number, b2: number) {
  return ((i - a) * (b2 - a2)) / (b - a) + a2;
}
function parseShadow(raw: any) {
  const values = raw.split(' ').map((val: any) => +val.replace('px', ''));
  return {
    x: values[0],
    y: values[1],
    blur: values[2],
    spread: values[3], // unused
  };
}
function onDepthChange(depth: number) {
  const s = parseShadow(androidDepth.penumbra[depth]);
  const y = s.y === 1 ? 1 : Math.floor(s.y * 0.5);
  return {
    elevation: depth + 1, // Android
    // iOS
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: y,
    },
    shadowOpacity: Number.parseFloat(interpolate(depth, 1, 24, 0.2, 0.6).toFixed(2)),
    shadowRadius: Number.parseFloat(interpolate(s.blur, 1, 38, 1, 16).toFixed(2)),
  };
}

class CAppView {
  private static _instance: CAppView;

  private constructor() {
    // ...
  }

  public static get Instance(): CAppView {
    /* istanbul ignore else */
    if (!this._instance) {
      this._instance = new this();
    }
    return CAppView._instance;
  }

  fontFamily: Font = 'GoogleSans-Regular';

  safeAreaInsets = {
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  };

  shadow = (depth: number = 10) => (depth >= 1 ? onDepthChange(depth - 1) : {});

  bottomNavigationBarHeight = 45;

  headerHeight = 50;

  bodyHeight =
    height -
    this.headerHeight -
    this.bottomNavigationBarHeight -
    this.safeAreaInsets.top -
    this.safeAreaInsets.bottom;

  bodyPaddingHorizontal = Number.parseFloat(ViewConfig.spacing.DEFAULT);

  bodyWidth = width - 2 * this.bodyPaddingHorizontal;

  roundedBorderRadius = Number.parseFloat(ViewConfig.borderRadius.DEFAULT);

  bottomSheetBorderRadius = Number.parseFloat(ViewConfig.borderRadius.card);

  itemMarginHorizontal = 10;

  screenWidth = width;

  screenHeight = height;

  isHorizontal = this.screenWidth > this.screenHeight;

  // Carousel
  sliderWidth = this.screenWidth;

  itemWidth = (this.isHorizontal ? 0.35 : 0.75) * this.screenWidth + this.itemMarginHorizontal * 2;

  itemHeight = 300;

  carouselBorderRadius = 8;

  onDimensionChange = (dimensions: { window: ScaledSize; screen: ScaledSize }) => {
    const { window } = dimensions;
    this.screenWidth = window.width;
    this.screenHeight = window.height;
    const isHorizontal = window.width > window.height;
    this.isHorizontal = isHorizontal;

    if (isHorizontal) {
      // Carousel
      this.itemWidth = 0.35 * window.width + this.itemMarginHorizontal * 2;
      this.sliderWidth = window.width;
    } else {
      this.itemWidth = 0.75 * window.width + this.itemMarginHorizontal * 2;
      this.sliderWidth = window.width;
    }
  };

  initSafeArea = (safeAreaInsets: { top: number; bottom: number; left: number; right: number }) => {
    const dimensions = Dimensions.get('window');
    this.safeAreaInsets = safeAreaInsets;
    this.bodyHeight =
      dimensions.height -
      this.headerHeight -
      this.bottomNavigationBarHeight -
      safeAreaInsets.top -
      safeAreaInsets.bottom;
  };

  getHeaderStyle = (theme: typeof DefaultColor) =>
    ({
      headerStyle: {
        backgroundColor: theme.colors.background,
      } as StyleProp<ViewStyle>,
      headerTitleStyle: {
        fontSize: 22,
        color: theme.colors.grey0,
        fontWeight: 'bold',
        fontFamily: AppView.fontFamily,
      } as StyleProp<TextStyle>,
      headerShadowVisible: false,
      headerBackTitleVisible: false,
      headerBackButtonMenuEnabled: false,
      headerTintColor: theme.colors.text,
    } as any);

  getQuickStyle = ({
    outline,
    clear,
    backgroundColor,
    textColor,
    borderColor,
  }: {
    outline?: boolean;
    clear?: boolean;
    backgroundColor: ColorValue;
    textColor: ColorValue;
    borderColor?: ColorValue;
  }) => {
    let quickStyles: StyleProp<ViewStyle> = {};
    let quickTextStyles: StyleProp<TextStyle> = {};
    if (outline) {
      quickStyles = {
        backgroundColor: 'transparent',
        borderColor: borderColor || backgroundColor,
        borderWidth: 1,
        elevation: 0,
      };
      quickTextStyles = {
        color: textColor || backgroundColor,
      };
    } else if (clear) {
      quickStyles = {
        backgroundColor: 'transparent',
        elevation: 0,
        borderRadius: 0,
      };
      quickTextStyles = { color: backgroundColor };
    } else {
      quickStyles = {
        backgroundColor,
      };
    }
    return { quickStyles, quickTextStyles };
  };

  getInputQuickStyle = ({
    outline,
    clear,
  }: {
    outline?: boolean;
    clear?: boolean;
    textColor: ColorValue;
  }) => {
    let quickStyles: StyleProp<ViewStyle> = {};
    if (outline) {
      quickStyles = {
        borderWidth: 1,
        elevation: 0,
      };
    } else if (clear) {
      quickStyles = {
        backgroundColor: 'transparent',
        borderRadius: 0,
        borderBottomWidth: 1,
        elevation: 0,
      };
    } else {
      quickStyles = {
        borderWidth: 1,
      };
    }
    return { quickStyles };
  };
}

const AppView = CAppView.Instance;
export default AppView;
