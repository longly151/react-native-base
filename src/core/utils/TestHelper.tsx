import React, { useLayoutEffect } from 'react';
import { render, RenderAPI } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import store, { persistor, createStore } from '@configs/store';
import { LoadingScreen } from '@app';
import { NavigationContainer } from '@react-navigation/native';
import Navigation from './Navigation';
import MockAdapter from 'axios-mock-adapter';
import { Axios } from '@utils/Api';

import DefaultColor from '@themes/DefaultColor';
import { RootState, useAppDispatch, useAppSelector } from './Redux';
import _ from 'lodash';

import { renderHook } from '@testing-library/react-hooks';

type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>;
    }
  : T;

/**
 * Login Handler
 */
export const successCredential = { email: 'admin@gmail.com', password: '123456' };
export const successLoginResponse = {
  data: {
    token:
      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwic3RhdHVzIjoiQUNUSVZFIiwicm9sZSI6IkFETUlOIiwicm9sZUlkIjoxLCJwZXJtaXNzaW9ucyI6WyJBTEwiXSwidGFibGUiOiJ1c2VycyIsImlkRm9yZWlnbktleSI6InVzZXJJZCIsImlhdCI6MTY1NTQ1NzU5OCwiZXhwIjoxNjYwNjQxNTk4fQ.ZA1dsxArHAjb60HzQgKbkinl49cf5OiDWASiE_Wr0SA',
    id: 1,
    email: 'admin@gmail.com',
    fullName: 'Admin',
    phone: '0327571918',
    avatar:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcRlUbAyS_643dq_B69jZAlPNW6_Xc7SLELY6SpRsc5OI2wHiiYG&usqp=CAU',
    role: {
      createdAt: '2021-04-19T05:58:23.094Z',
      updatedAt: '2021-04-19T05:58:23.094Z',
      deletedAt: null,
      id: 1,
      name: 'ADMIN',
      description: 'ADMIN',
      permissions: [
        {
          id: 1,
          name: 'ALL',
          description: null,
        },
      ],
    },
  },
};

export const failCredential = { email: 'client@gmail.com', password: '123456' };
export const failLoginResponse = {
  statusCode: 401,
  message: [
    {
      code: 'INCORRECT_EMAIL_PASSWORD',
      description: 'Incorrect email or password',
    },
  ],
  error: 'Unauthorized',
};

const auth = {
  loading: false,
  data: successLoginResponse.data,
  error: null,
};
const withAuthStore = createStore({ auth });

/**
 * Wrapper handler
 */

export type OtherRenderOptions = {
  preloadedState?: DeepPartial<RootState>;
  withAuth?: boolean;
  disableNavigationRef?: boolean;
};

const AppContainer = ({
  children,
  disableNavigationRef,
}: {
  children: any;
  disableNavigationRef?: boolean;
}) => {
  const [loginData] = useAppSelector(state => [state.auth.data]);

  const dispatch = useAppDispatch();

  useLayoutEffect(() => {
    const initApp = () => {
      if (loginData?.token) {
        global.token = loginData.token;
      }
      global.fn = {} as any;
      global.dispatch = dispatch;
    };
    initApp();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return (
    <NavigationContainer
      ref={disableNavigationRef ? null : Navigation.navigationRef}
      theme={DefaultColor}
    >
      {children}
    </NavigationContainer>
  );
};

const AppProviders = ({
  children,
  preloadedState,
  withAuth,
  disableNavigationRef,
}: {
  children: any;
} & OtherRenderOptions) => {
  const mergedPreloadedState: DeepPartial<RootState> = withAuth
    ? _.merge({ auth }, preloadedState as any)
    : preloadedState;

  return (
    // Create new store if preloadedState !== undefined
    <Provider
      store={preloadedState ? createStore(mergedPreloadedState) : withAuth ? withAuthStore : store}
    >
      <PersistGate loading={LoadingScreen()} persistor={persistor}>
        <AppContainer disableNavigationRef={disableNavigationRef}>{children}</AppContainer>
      </PersistGate>
    </Provider>
  );
};

const customRender = (
  component: React.ReactElement<any>,
  // options?: RenderOptions & OtherRenderOptions,
  options?: OtherRenderOptions,
): RenderAPI => {
  const CustomWrapper = options?.wrapper || React.Fragment;
  const Wrapper = options?.wrapper
    ? ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={options.disableNavigationRef}
        >
          <CustomWrapper>{children}</CustomWrapper>
        </AppProviders>
      )
    : ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={options?.disableNavigationRef}
        >
          {children}
        </AppProviders>
      );

  return render(component, { ...options, wrapper: Wrapper });
};

const customRenderHook: typeof renderHook = (callback, options) => {
  const disableNavigationRef =
    typeof options?.disableNavigationRef === 'undefined' ? true : options?.disableNavigationRef;
  const CustomWrapper: any = options?.wrapper || React.Fragment;
  const Wrapper = options?.wrapper
    ? ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={disableNavigationRef}
        >
          <CustomWrapper>{children}</CustomWrapper>
        </AppProviders>
      )
    : ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={disableNavigationRef}
        >
          {children}
        </AppProviders>
      );

  return renderHook(callback, { ...options, wrapper: Wrapper });
};

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };

export { customRenderHook as renderHook };

const mockAxios = new MockAdapter(Axios, { delayResponse: 200 });

export { mockAxios };
