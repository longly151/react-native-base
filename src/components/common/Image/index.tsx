import {
  StyleProp,
  Image as RNImage,
  Animated,
  TouchableOpacity,
  ImageURISource,
  ImageRequireSource,
  ViewStyle,
  StyleSheet,
  View,
} from 'react-native';
import React, { useEffect, useImperativeHandle, useRef, useState } from 'react';
import withTailwind from '@components/hoc/withTailwind';
import ImageSource from '@images';
import { useTheme } from '@react-navigation/native';
import Loading from '../Loading';
import _ from 'lodash';
import AppView from '@utils/AppView';
import ImageViewerModal from './components/ImageViewerModal';
import ImageViewer from 'react-native-image-zoom-viewer';
import FastImage, { FastImageProps, ImageStyle, OnLoadEvent } from 'react-native-fast-image';

export interface ImageProps extends FastImageProps {
  className?: string;
  autoHeight?: boolean;
  autoWidth?: boolean;
  viewEnable?: boolean;
  multipleSources?: Array<ImageURISource | ImageRequireSource>;
  disablePlaceholder?: boolean;
  sharp?: boolean;
  width?: number | string;
  height?: number | string;
  maxWidth?: number;
  maxHeight?: number;
}

const containerStyleKeys = [
  'alignSelf',
  'margin',
  'marginLeft',
  'marginRight',
  'marginTop',
  'marginBottom',
  'marginVertical',
  'marginHorizontal',
  'position',
  'borderColor',
  'borderTopColor',
  'borderBottomColor',
  'borderLeftColor',
  'borderRightColor',
];

const Image: React.FC<ImageProps> = React.forwardRef((props, ref: any) => {
  const imageViewerModalRef = useRef<ImageViewer>(null);

  const theme = useTheme();

  const {
    style: rawStyleProp,
    width: widthProp,
    height: heightProp,
    maxWidth,
    maxHeight,
    autoHeight,
    autoWidth,
    viewEnable,
    multipleSources,
    source,
    disablePlaceholder,
    sharp,
    ...otherProps
  } = props;
  const styleProp: StyleProp<ImageStyle> = StyleSheet.flatten(rawStyleProp);

  // @ts-ignore
  const isLocalImage = !source?.uri;

  const [isModalVisible, setModalVisible] = useState(false);
  const [loading, setLoading] = useState(!isLocalImage);

  const localDimensions: any = isLocalImage ? RNImage.resolveAssetSource(source as any) : null;

  const [width, setWidth] = useState(
    widthProp || (styleProp?.width as number | undefined) || AppView.bodyWidth,
  );
  const [height, setHeight] = useState(
    heightProp ||
      (typeof width === 'number'
        ? (styleProp?.height as number | undefined) ||
          (localDimensions
            ? (localDimensions.height / localDimensions.width) * width
            : AppView.bodyWidth)
        : undefined),
  );

  useEffect(() => {
    if (maxWidth && typeof width === 'number' && width > maxWidth) {
      setWidth(maxWidth);
    }
    if (maxHeight && typeof height === 'number' && height > maxHeight) {
      setHeight(maxHeight);
    }
  }, [maxWidth, maxHeight, width, height]);

  const [imageDimensions, setImageDimensions] = useState({ width, height } as {
    width: number;
    height: number;
  });

  const fadeInAnim = useRef(new Animated.Value(0)).current;
  const fadeOutAnim = useRef(new Animated.Value(1)).current;

  const onLoad = (e: OnLoadEvent) => {
    const {
      nativeEvent: { height: imageHeight, width: imageWidth },
    } = e;

    const ratio = imageHeight / imageWidth;

    if (
      !autoWidth &&
      autoHeight &&
      !heightProp &&
      !styleProp?.height &&
      typeof width === 'number'
    ) {
      if (height !== ratio * width) {
        setHeight(ratio * width);
      }
    }
    if (autoWidth && !widthProp && !styleProp.width && typeof height === 'number') {
      if (width !== height / ratio) {
        setWidth(height / ratio);
      }
    }
    setImageDimensions({ width: imageWidth, height: imageHeight });
    setLoading(false);

    Animated.timing(fadeInAnim, {
      toValue: 1,
      duration: 500,
      useNativeDriver: true,
    }).start();
    Animated.timing(fadeOutAnim, {
      toValue: 0,
      duration: 250,
      useNativeDriver: true,
    }).start();
    props.onLoad?.(e);
  };

  const style: StyleProp<ImageStyle> = StyleSheet.flatten([
    {
      width,
      height,
      borderRadius: sharp ? 0 : AppView.roundedBorderRadius,
    },
    styleProp,
  ]);
  const containerStyle: StyleProp<ViewStyle> = {
    ...(_.pick(style, containerStyleKeys) as any),
    width,
    height,
    borderRadius: (style as any).borderRadius,
  };
  const imageStyle: StyleProp<ImageStyle> = _.omit(style, containerStyleKeys) as any;

  const onViewImage = () => {
    setModalVisible(true);
  };

  const renderPlaceholder = () => {
    if (!disablePlaceholder && loading) {
      return (
        <Animated.View style={{ position: 'absolute', opacity: fadeOutAnim }}>
          <FastImage
            testID="PlaceholderImage"
            source={ImageSource.placeholderImage}
            style={[
              imageStyle,
              {
                tintColor: theme.dark ? theme.colors.grey1 : theme.colors.grey3,
              },
            ]}
            resizeMode="contain"
          />
        </Animated.View>
      );
    }
    if (loading) {
      return (
        <Loading
          testID="Loading"
          style={{ position: 'absolute', alignSelf: 'center', height: '100%' }}
        />
      );
    }
    return null;
  };

  const renderImage = (customProps?: any) => {
    const isFullScreen = !!customProps;
    const ratio =
      typeof width === 'number' && typeof height === 'number'
        ? (height * 1.0) / width
        : typeof width === 'string' && typeof height === 'string'
        ? Number.parseFloat(height) / Number.parseFloat(width)
        : 1;

    const viewModeHeight = customProps?.style?.height || ratio * AppView.screenWidth;
    const imgSource = customProps?.source || source;
    const imgStyle: any = StyleSheet.flatten([
      imageStyle,
      {
        width: isModalVisible ? AppView.screenWidth : width,
        height: isModalVisible ? viewModeHeight : height,
        borderRadius: isModalVisible ? 0 : style.borderRadius,
        borderTopLeftRadius: isModalVisible ? 0 : style.borderTopLeftRadius,
        borderTopRightRadius: isModalVisible ? 0 : style.borderTopRightRadius,
        borderBottomLeftRadius: isModalVisible ? 0 : style.borderBottomLeftRadius,
        borderBottomRightRadius: isModalVisible ? 0 : style.borderBottomRightRadius,
      },
    ]);

    return (
      <Animated.View style={{ opacity: isModalVisible || isLocalImage ? 1 : fadeInAnim }}>
        <FastImage
          testID={`Image${isFullScreen ? 'FullScreen' : ''}`}
          {...otherProps}
          key={isModalVisible.toString()}
          style={imgStyle}
          onLoad={onLoad}
          source={imgSource}
        />
      </Animated.View>
    );
  };

  useImperativeHandle(ref, () => ({
    imageViewerModalRef,
  }));

  return (
    <View>
      <TouchableOpacity
        testID="ContainerTouchableOpacity"
        disabled={!viewEnable || !source}
        onPress={viewEnable ? onViewImage : undefined}
        style={containerStyle}
      >
        {renderImage()}
        {renderPlaceholder()}
      </TouchableOpacity>
      <ImageViewerModal
        ref={imageViewerModalRef}
        isModalVisible={isModalVisible}
        setModalVisible={setModalVisible}
        source={source}
        multipleSources={multipleSources}
        imageDimensions={imageDimensions}
        renderImage={renderImage}
      />
    </View>
  );
});

Image.defaultProps = {
  autoHeight: true,
};

export interface ImageType extends RNImage {
  imageViewerModalRef: React.RefObject<ImageViewer>;
}

export default withTailwind(Image) as React.FC<ImageProps & { ref?: React.LegacyRef<RNImage> }>;
