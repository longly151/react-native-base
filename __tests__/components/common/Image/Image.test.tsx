import React from 'react';
import RN, { Image as RNImage } from 'react-native';
import { act, fireEvent, render, waitFor } from '@utils/TestHelper';
import Image, { ImageType } from '@components/common/Image/index';
import ImageSource from '@images';
import AppView from '@utils/AppView';
import DefaultColor from '@themes/DefaultColor';
import DarkColor from '@themes/DarkColor';

global.console = {
  ...console,
  error: jest.fn(),
};

jest.spyOn(RN.Animated, 'FlatList', 'get').mockImplementation(() => RN.FlatList);

let mockTheme = DefaultColor;
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useTheme: jest.fn().mockImplementation(() => mockTheme),
}));

const localSource = ImageSource.iconAppIOS;
const remoteSource = {
  uri: 'https://images.unsplash.com/photo-1518005020951-eccb494ad742?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1500&q=80',
};
const imageWidth = 100;
const imageHeight = imageWidth * 2;

RNImage.resolveAssetSource = jest.fn().mockReturnValue({ width: imageWidth, height: imageHeight });

test('should render correctly', () => {
  const { toJSON, getByTestId } = render(<Image source={localSource} />);
  expect(toJSON()).toMatchSnapshot();
  expect(getByTestId('Image')).toBeTruthy();
});

test('should have working props with local image', async () => {
  /**
   * Local Image
   */
  const { getByTestId } = render(
    <Image
      source={localSource}
      className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500"
    />,
  );

  expect(getByTestId('ContainerTouchableOpacity')).toHaveStyle({
    alignSelf: 'center',
    marginLeft: 4,
    marginRight: 4,
    marginTop: 4,
    marginBottom: 4,
    borderLeftColor: 'rgba(3, 105, 161, 1)',
    borderRightColor: 'rgba(3, 105, 161, 1)',
    borderTopColor: 'rgba(3, 105, 161, 1)',
    borderBottomColor: 'rgba(3, 105, 161, 1)',
  });

  expect(getByTestId('Image')).toHaveStyle({
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderBottomLeftRadius: 4,
    borderBottomRightRadius: 4,
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4,
    borderLeftWidth: 1,
    borderRightWidth: 1,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    backgroundColor: 'rgba(14, 165, 233, 1)',
    paddingLeft: 4,
    paddingRight: 4,
    paddingTop: 4,
    paddingBottom: 4,

    // Image Dimensions
    width: AppView.bodyWidth,
    height: AppView.bodyWidth * 2,
  });
});

test('should have working props with remote image', async () => {
  /**
   * Remote Image (with width = screenWidth & autoHeight = true in default)
   */
  const { getByTestId, findByTestId, queryByTestId } = render(
    <Image source={remoteSource} viewEnable sharp />,
  );

  expect(getByTestId('Image')).toHaveStyle({
    borderRadius: 0, // Sharp Image
    width: AppView.bodyWidth,
    height: AppView.bodyWidth,
  });

  expect(getByTestId('PlaceholderImage')).toHaveStyle({
    tintColor: mockTheme.colors.grey3,
  });

  fireEvent(getByTestId('Image'), 'onLoad', {
    nativeEvent: {
      source: { width: imageWidth, height: imageHeight },
    },
  });

  expect(getByTestId('Image')).toHaveStyle({
    width: AppView.bodyWidth,
    height: AppView.bodyWidth * 2,
  });
  expect(queryByTestId('PlaceholderImage')).toBeFalsy();

  // Open Image in FullScreen View
  fireEvent.press(getByTestId('ContainerTouchableOpacity'));
  const ImageViewerModal = await findByTestId('ImageViewerModal');

  expect(ImageViewerModal).toHaveProp('visible', true);
  expect(getByTestId('ImageFullScreen')).toHaveStyle({
    width: AppView.screenWidth,
    height: AppView.screenWidth * 2,
  });

  /**
   * Remote Image (with autoWidth = true)
   */
  const { getByTestId: getByTestIdAutoWidth } = render(
    <Image source={remoteSource} viewEnable autoWidth />,
  );

  expect(getByTestIdAutoWidth('Image')).toHaveStyle({
    height: AppView.bodyWidth,
    width: AppView.bodyWidth,
  });

  fireEvent(getByTestIdAutoWidth('Image'), 'onLoad', {
    nativeEvent: {
      source: { width: imageWidth, height: imageHeight },
    },
  });

  expect(getByTestIdAutoWidth('Image')).toHaveStyle({
    height: AppView.bodyWidth,
    width: AppView.bodyWidth / 2,
  });

  /**
   * Remote Image (with custom width, height)
   */
  const { getByTestId: getByTestIdCustomDimensions } = render(
    <Image source={remoteSource} width={500} height={1000} />,
  );
  expect(getByTestIdCustomDimensions('Image')).toHaveStyle({
    width: 500,
    height: 1000,
  });

  /**
   * Show Loading (disablePlaceholder === true)
   */
  const { findByTestId: findByTestIdWithLoading } = render(
    <Image source={remoteSource} viewEnable disablePlaceholder width={200} />,
  );

  expect(await findByTestIdWithLoading('Loading')).toBeTruthy();

  /**
   * DarkColor
   */
  mockTheme = DarkColor;
  const { getByTestId: getByTestIdDark } = render(<Image source={remoteSource} />);

  expect(getByTestIdDark('PlaceholderImage')).toHaveStyle({
    tintColor: mockTheme.colors.grey1,
  });
});

test('should render fullscreen view correctly', async () => {
  const imageRef: React.RefObject<ImageType> = React.createRef();

  const { getByTestId: getByTestIdImage } = render(
    <Image
      ref={imageRef}
      source={localSource}
      multipleSources={[remoteSource, localSource]}
      viewEnable
    />,
  );

  // Open Image in FullScreen View
  fireEvent.press(getByTestIdImage('ContainerTouchableOpacity'));

  expect(getByTestIdImage('ImageViewerModal')).toHaveProp('visible', true);
  expect(imageRef.current?.imageViewerModalRef.current?.props.index).toBe(1); // localSource index

  // Close Image FullScreen View
  act(() => {
    imageRef.current?.imageViewerModalRef.current?.handleCancel();
  });

  await waitFor(() => expect(getByTestIdImage('ImageViewerModal')).toHaveProp('visible', false));
});
