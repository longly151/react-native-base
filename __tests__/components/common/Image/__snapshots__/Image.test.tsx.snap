// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<View>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderRadius": 8,
        "height": 1420,
        "opacity": 1,
        "width": 710,
      }
    }
    testID="ContainerTouchableOpacity"
  >
    <Image
      collapsable={false}
      onLoad={[Function]}
      source={
        Object {
          "testUri": "../../../src/assets/images/app/iconIos.png",
        }
      }
      style={
        Object {
          "borderBottomLeftRadius": undefined,
          "borderBottomRightRadius": undefined,
          "borderRadius": 8,
          "borderTopLeftRadius": undefined,
          "borderTopRightRadius": undefined,
          "height": 1420,
          "opacity": 1,
          "width": 710,
        }
      }
      testID="Image"
    />
  </View>
  <Modal
    animationType="fade"
    hardwareAccelerated={true}
    presentationStyle="overFullScreen"
    statusBarTranslucent={true}
    testID="ImageViewerModal"
    transparent={true}
    visible={false}
  />
</View>
`;
